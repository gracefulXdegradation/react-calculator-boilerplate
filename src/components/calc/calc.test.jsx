import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import Calc from './index';

const getScreen = wrapper => wrapper.find('[data-role="screen"]');
const clickButton = (button, wrapper) => wrapper.find(`[data-role="button_${button}"]`).simulate('click');
const pressKey = key => act(() => {
    document.body.dispatchEvent(new KeyboardEvent('keypress', { key }));
});
const pressKeys = keys => keys.split('').forEach(key => pressKey(key));
const clickButtons = (buttons, wrapper) => buttons.split('').forEach(button => clickButton(button, wrapper));

describe('App', () => {
    let consoleSpy;
    let wrapper;

    beforeEach(() => {
        wrapper = mount(<Calc />);
    });

    beforeAll(() => {
        consoleSpy = jest.spyOn(console, 'error');
    });

    it('renders the Calc component and matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    // it('renders logo', () => {
    //     expect(wrapper.find('img')).toHaveLength(1);
    // });

    it('shows no errors on console', () => {
        expect(consoleSpy).not.toHaveBeenCalled();
    });

    it('shows 0 as init value', () => {
        const screen = getScreen(wrapper);

        expect(screen).toHaveLength(1);
        expect(screen.props().value).toEqual('0');
    });

    it('handles number clicks correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('123', wrapper);
        expect(screen.getDOMNode().value).toEqual('123');
    });

    it('handles decimal numbers input correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('0.002300', wrapper);
        expect(screen.getDOMNode().value).toEqual('0.002300');
    });

    it('allows max one decimal separator', () => {
        const screen = getScreen(wrapper);

        clickButton('.', wrapper);
        expect(screen.getDOMNode().value).toEqual('0.');
        clickButtons('123.4', wrapper);
        expect(screen.getDOMNode().value).toEqual('0.1234');
    });

    it('limits number of digits in the input', () => {
        const screen = getScreen(wrapper);

        clickButtons('123456.78901234567890', wrapper);
        expect(screen.getDOMNode().value).toEqual('123 456.789');
    });

    it('groups digits in integer part correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('1234', wrapper);
        expect(screen.getDOMNode().value).toEqual('1 234');
        clickButton('5', wrapper);
        expect(screen.getDOMNode().value).toEqual('12 345');
        clickButtons('.6789', wrapper);
        expect(screen.getDOMNode().value).toEqual('12 345.6789');
    });

    it('starts new input after operation button click', () => {
        const screen = getScreen(wrapper);

        clickButtons('1234+.5678', wrapper);
        expect(screen.getDOMNode().value).toEqual('0.5678');
        clickButtons('=45.6', wrapper);
        expect(screen.getDOMNode().value).toEqual('45.6');
    });

    it('adds correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('123.4+567.8=', wrapper);
        expect(screen.getDOMNode().value).toEqual('691.2');
    });

    it('subtracts correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('123.4-567.8=', wrapper);
        expect(screen.getDOMNode().value).toEqual('-444.4');
    });

    it('chains operations correctly', () => {
        const screen = getScreen(wrapper);

        clickButtons('666-999+777+13-69=', wrapper);
        expect(screen.getDOMNode().value).toEqual('388');
        clickButtons('+112=', wrapper);
        expect(screen.getDOMNode().value).toEqual('500');
    });

    it('trims trailing zeros correctly: int', () => {
        const screen = getScreen(wrapper);

        clickButtons('.1+1.9=', wrapper);
        expect(screen.getDOMNode().value).toEqual('2');
    });

    it('trims trailing zeros correctly: float', () => {
        const screen = getScreen(wrapper);

        clickButtons('.002+1.898=', wrapper);
        expect(screen.getDOMNode().value).toEqual('1.9');
    });

    it('handles key press correctly', () => {
        const screen = getScreen(wrapper);

        pressKeys('123.4+56.7-89.0=');
        expect(screen.getDOMNode().value).toEqual('91.1');
    });
});
