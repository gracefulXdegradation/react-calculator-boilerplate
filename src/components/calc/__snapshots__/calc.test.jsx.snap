// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the Calc component and matches snapshot 1`] = `
<Calc>
  <form>
    <div>
      <img
        alt="DCMN logo"
        className="DCMN-logo"
        src="https://dcmn.com/assets/dcmn-logo-redesign-black.svg"
      />
      <input
        data-role="screen"
        readOnly={true}
        type="text"
        value="0"
      />
    </div>
    <div
      className="keyboard"
      onMouseLeave={[Function]}
    >
      <div
        className="pad hidden"
      >
        <div />
      </div>
      <div
        className="numboard"
      >
        <Button
          caption="1"
          className=""
          highlighted={false}
          key="button_1"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_1"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              1
            </span>
          </button>
        </Button>
        <Button
          caption="2"
          className=""
          highlighted={false}
          key="button_2"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_2"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              2
            </span>
          </button>
        </Button>
        <Button
          caption="3"
          className=""
          highlighted={false}
          key="button_3"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_3"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              3
            </span>
          </button>
        </Button>
        <Button
          caption="4"
          className=""
          highlighted={false}
          key="button_4"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_4"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              4
            </span>
          </button>
        </Button>
        <Button
          caption="5"
          className=""
          highlighted={false}
          key="button_5"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_5"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              5
            </span>
          </button>
        </Button>
        <Button
          caption="6"
          className=""
          highlighted={false}
          key="button_6"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_6"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              6
            </span>
          </button>
        </Button>
        <Button
          caption="7"
          className=""
          highlighted={false}
          key="button_7"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_7"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              7
            </span>
          </button>
        </Button>
        <Button
          caption="8"
          className=""
          highlighted={false}
          key="button_8"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_8"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              8
            </span>
          </button>
        </Button>
        <Button
          caption="9"
          className=""
          highlighted={false}
          key="button_9"
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_9"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              9
            </span>
          </button>
        </Button>
      </div>
      <div
        className="opsboard"
      >
        <Button
          caption="+"
          className=""
          highlighted={false}
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_+"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              +
            </span>
          </button>
        </Button>
        <Button
          caption="-"
          className=""
          highlighted={false}
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_-"
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              -
            </span>
          </button>
        </Button>
        <Button
          caption="="
          className=""
          highlighted={false}
          onClick={[Function]}
          onHover={[Function]}
        >
          <button
            className=""
            data-role="button_="
            onClick={[Function]}
            onMouseEnter={[Function]}
            type="button"
          >
            <span>
              =
            </span>
          </button>
        </Button>
      </div>
      <Button
        caption="."
        className=""
        highlighted={false}
        onClick={[Function]}
        onHover={[Function]}
      >
        <button
          className=""
          data-role="button_."
          onClick={[Function]}
          onMouseEnter={[Function]}
          type="button"
        >
          <span>
            .
          </span>
        </button>
      </Button>
      <Button
        caption="0"
        className=""
        highlighted={false}
        onClick={[Function]}
        onHover={[Function]}
      >
        <button
          className=""
          data-role="button_0"
          onClick={[Function]}
          onMouseEnter={[Function]}
          type="button"
        >
          <span>
            0
          </span>
        </button>
      </Button>
      <Button
        caption="AC"
        className="ac"
        highlighted={false}
        onClick={[Function]}
        onHover={[Function]}
      >
        <button
          className="ac"
          data-role="button_ac"
          onClick={[Function]}
          onMouseEnter={[Function]}
          type="button"
        >
          <span>
            AC
          </span>
        </button>
      </Button>
    </div>
  </form>
</Calc>
`;
